File: lib/spendable_web/utils/budget_card.ex
defmodule SpendableWeb.Utils.BudgetCard do
  @moduledoc """
  What a budget reads as on the budgets screen: the number to show, what to call it, how far
  along the bar is, and the line underneath.

  Extracted from the LiveView rather than left private because the iOS client has to reach the
  same six answers, and `shared/budget_cards.json` drives a table test on both sides. Colours
  stay with each client - `bar` says which of the four bars this is, not what it looks like.

  The month is passed as one map of what moved - `spent` and `received` - because which of them a
  card reads depends on what kind of budget it is, and a budget that spends never receives.
  """

  import Spendable.Utils

  alias Spendable.Budgets.Schemas.Budget

  # A past month is a record of what moved, so a balance read now says nothing about it.
  def build_budget_card(%Budget{type: :income}, %{received: received}, false = _current_month) do
    %{amount: received, label: "EARNED", percent: nil, bar: nil, footer: nil}
  end

  def build_budget_card(_budget, %{spent: spent}, false = _current_month_is_selected) do
    %{amount: spent, label: "SPENT", percent: nil, bar: nil, footer: nil}
  end

  def build_budget_card(%Budget{type: :tracking, budgeted_amount: nil}, %{spent: spent}, _current) do
    %{amount: spent, label: "SPENT", percent: nil, bar: nil, footer: nil}
  end

  def build_budget_card(%Budget{type: :tracking} = budget, %{spent: spent}, _current) do
    %{
      amount: spent,
      label: "SPENT",
      percent: percent(spent, budget.budgeted_amount),
      bar: spending_bar(spent, budget.budgeted_amount),
      footer: "#{format_currency(spent)} of #{format_currency(budget.budgeted_amount)} spent"
    }
  end

  def build_budget_card(%Budget{type: :income, budgeted_amount: nil}, %{received: received}, _current) do
    %{amount: received, label: "EARNED", percent: nil, bar: nil, footer: nil}
  end

  # Money in is the point here, so there is no bar to be the wrong side of: it fills as the month
  # earns and the footer says how far along that is.
  def build_budget_card(%Budget{type: :income} = budget, %{received: received}, _current) do
    %{
      amount: received,
      label: "EARNED",
      percent: percent(received, budget.budgeted_amount),
      bar: "income",
      footer: "#{format_currency(received)} of #{format_currency(budget.budgeted_amount)} received"
    }
  end

  def build_budget_card(%Budget{type: :envelope, funding_amount: nil} = budget, _month, _current) do
    held(budget)
  end

  # An envelope has one amount: what a month puts in, which is also what its spending is read
  # against. There is nothing to say about funding separately because they are the same figure.
  def build_budget_card(%Budget{type: :envelope} = budget, %{spent: spent}, _current) do
    %{
      held(budget)
      | percent: percent(spent, budget.funding_amount),
        bar: spending_bar(spent, budget.funding_amount),
        footer: "#{format_currency(spent)} of #{format_currency(budget.funding_amount)} spent"
    }
  end

  def build_budget_card(%Budget{type: :goal, budgeted_amount: nil} = budget, _month, _current) do
    %{amount: budget.balance, label: "SAVED", percent: nil, bar: nil, footer: "No goal set"}
  end

  def build_budget_card(%Budget{type: :goal} = budget, _month, _current_month_is_selected) do
    saved_line =
      "#{format_currency(budget.balance)} of #{format_currency(budget.budgeted_amount)} saved"

    %{
      amount: Decimal.sub(budget.budgeted_amount, budget.balance),
      label: "TO GO",
      percent: percent(budget.balance, budget.budgeted_amount),
      bar: "goal",
      footer: suffix_monthly(saved_line, budget.funding_amount)
    }
  end

  # An envelope in the hole is not holding a negative amount, it is short by a positive one - the
  # same way a goal counts what is still TO GO rather than a negative saving. Clients colour the
  # label, since the figure no longer carries a minus sign to key off.
  defp held(%Budget{balance: balance} = budget) do
    if Decimal.negative?(balance) do
      %{amount: Decimal.abs(balance), label: "OVERSPENT", percent: nil, bar: nil, footer: nil}
    else
      %{amount: budget.balance, label: "REMAINING", percent: nil, bar: nil, footer: nil}
    end
  end

  defp spending_bar(spent, budgeted_amount) do
    if Decimal.compare(spent, budgeted_amount) == :gt, do: "over", else: "under"
  end

  defp suffix_monthly(line, nil), do: line
  defp suffix_monthly(line, funding_amount), do: "#{line} · #{format_currency(funding_amount)}/mo"

  defp percent(_part, %Decimal{coef: 0}), do: 0.0

  defp percent(part, whole) do
    part
    |> Decimal.div(whole)
    |> Decimal.mult(100)
    |> Decimal.to_float()
    |> max(0.0)
    |> min(100.0)
    |> Float.round(1)
  end
end

File: mobile/lib/budgets/budget_card.dart
import 'package:decimal/decimal.dart';
import 'package:spendable_api/spendable_api.dart';

import '../money.dart';

/// Which of the four bars a card draws. Colours are the screen's business.
enum CardBar { under, over, goal, income }

/// What one month moved through a budget. Which figure a card reads depends on what kind of
/// budget it is, and a budget that spends never receives.
class BudgetMonth {
  const BudgetMonth({required this.spent, required this.received, required this.funded});

  final Decimal spent;
  final Decimal received;
  final Decimal funded;
}

/// What a budget reads as, mirroring `SpendableWeb.Utils.BudgetCard`. Both are driven by
/// shared/budget_cards.json, which is the only thing keeping them from drifting apart.
class BudgetCard {
  const BudgetCard({required this.amount, required this.label, this.percent, this.bar, this.footer});

  factory BudgetCard.build({
    required Budget budget,
    required BudgetMonth month,
    required bool currentMonth,
  }) {
    final income = budget.type == BudgetTypeEnum.income;
    final budgeted = budget.budgetedAmount == null ? null : money(budget.budgetedAmount!);
    final spent = month.spent;

    // A past month is a record of what moved, so a balance read now says nothing about it.
    if (!currentMonth) {
      return income
          ? BudgetCard(amount: month.received, label: 'EARNED')
          : BudgetCard(amount: spent, label: 'SPENT');
    }

    if (budget.type == BudgetTypeEnum.tracking) {
      if (budgeted == null) return BudgetCard(amount: spent, label: 'SPENT');

      return BudgetCard(
        amount: spent,
        label: 'SPENT',
        percent: _percent(spent, budgeted),
        bar: spent > budgeted ? CardBar.over : CardBar.under,
        footer: '${formatCurrency(spent)} of ${formatCurrency(budgeted)} spent',
      );
    }

    // Money in is the point here, so there is no bar to be the wrong side of: it fills as the
    // month earns and the footer says how far along that is.
    if (income) {
      final received = month.received;

      if (budgeted == null) return BudgetCard(amount: received, label: 'EARNED');

      return BudgetCard(
        amount: received,
        label: 'EARNED',
        percent: _percent(received, budgeted),
        bar: CardBar.income,
        footer: '${formatCurrency(received)} of ${formatCurrency(budgeted)} received',
      );
    }

    final goal = budget.type == BudgetTypeEnum.goal;
    final balance = money(budget.balance);
    final fundsItself = budget.fundingAmount == null ? null : money(budget.fundingAmount!);
    final funded = month.funded;

    if (budgeted == null) {
      if (goal) return BudgetCard(amount: balance, label: 'SAVED', footer: 'No goal set');

      final held = _held(balance);

      return BudgetCard(amount: held.amount, label: held.label, footer: _fundedFooter(funded));
    }

    if (goal) {
      final saved = '${formatCurrency(balance)} of ${formatCurrency(budgeted)} saved';

      return BudgetCard(
        amount: budgeted - balance,
        label: 'TO GO',
        percent: _percent(balance, budgeted),
        bar: CardBar.goal,
        footer: fundsItself == null ? saved : '$saved · ${formatCurrency(fundsItself)}/mo',
      );
    }

    final spend = '${formatCurrency(spent)} of ${formatCurrency(budgeted)} spent';
    final held = _held(balance);

    return BudgetCard(
      amount: held.amount,
      label: held.label,
      percent: _percent(spent, budgeted),
      bar: spent > budgeted ? CardBar.over : CardBar.under,
      footer: funded == Decimal.zero ? spend : '${formatCurrency(funded)} funded · $spend',
    );
  }

  final Decimal amount;
  final String label;
  final double? percent;
  final CardBar? bar;
  final String? footer;

  /// An envelope in the hole is not holding a negative amount, it is short by a positive one -
  /// the same way a goal counts what is still TO GO rather than a negative saving. The screen
  /// colours the label, since the figure no longer carries a minus sign to key off.
  static ({Decimal amount, String label}) _held(Decimal balance) => balance < Decimal.zero
      ? (amount: -balance, label: 'OVERSPENT')
      : (amount: balance, label: 'LEFT');

  /// What the month put in is only worth saying when a month put something in, so a budget the
  /// user fills by hand reads exactly as it did before funding existed.
  static String? _fundedFooter(Decimal funded) =>
      funded == Decimal.zero ? null : '${formatCurrency(funded)} funded';

  static double _percent(Decimal part, Decimal whole) {
    if (whole == Decimal.zero) return 0;

    final value = (part / whole).toDecimal(scaleOnInfinitePrecision: 28).toDouble() * 100;

    return double.parse(value.clamp(0, 100).toStringAsFixed(1));
  }
}
